File: app/src/main/java/com/movies/popular/popularmovies/presentation/common/adapter/LoadingAdapter.java
package com.movies.popular.popularmovies.presentation.common.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.movies.popular.popularmovies.R;
import com.movies.popular.popularmovies.databinding.ItemLoadingBinding;

/**
 * Created with Android Studio.
 * User: Sasha Shcherbinin
 * Date: 9/30/17
 * Time: 11:07 AM
 */

public class LoadingAdapter extends RecyclerViewAdapterWrapper {

    private static final int ITEM_TYPE_LOADING = -1000;

    private final RecyclerView.AdapterDataObserver observer;

    private boolean loading = false;

    private boolean observerAdded;

    public LoadingAdapter(RecyclerView.Adapter adapter) {
        //noinspection unchecked
        super(adapter);
        observer = new RecyclerView.AdapterDataObserver() {

            public void onChanged() {
                notifyDataSetChanged();
            }

            public void onItemRangeChanged(int positionStart, int itemCount) {
                notifyItemRangeChanged(positionStart, itemCount);
            }

            public void onItemRangeInserted(int positionStart, int itemCount) {
                if (wrapped.getItemCount() == itemCount) {
                    notifyDataSetChanged();
                } else {
                    notifyItemRangeInserted(positionStart, itemCount);
                }
            }

            public void onItemRangeRemoved(int positionStart, int itemCount) {
                if (positionStart == 0 && itemCount >= wrapped.getItemCount()) {
                    notifyDataSetChanged();
                } else {
                    notifyItemRangeRemoved(positionStart, itemCount);
                }
            }

            public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
                notifyItemMoved(fromPosition, toPosition);
            }
        };
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == ITEM_TYPE_LOADING) {
            ItemLoadingBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),
                    R.layout.item_loading, parent, false);
            return new LoadingViewHolder(binding);
        } else {
            return wrapped.onCreateViewHolder(parent, viewType);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (!(holder instanceof LoadingViewHolder)) {
            wrapped.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (isLoadingPosition(position)) {
            return ITEM_TYPE_LOADING;
        } else {
            return wrapped.getItemViewType(position);
        }
    }

    @Override
    public int getItemCount() {
        int itemCount = wrapped.getItemCount();
        return loading ? itemCount + 1 : itemCount;
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        if (!(holder instanceof LoadingViewHolder)) {
            wrapped.onViewRecycled(holder);
        }
    }

    @Override
    public void registerAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        super.registerAdapterDataObserver(observer);
        if (!observerAdded) {
            this.wrapped.registerAdapterDataObserver(this.observer);
            observerAdded = true;
        }
    }

    @Override
    public void unregisterAdapterDataObserver(@NonNull RecyclerView.AdapterDataObserver observer) {
        super.unregisterAdapterDataObserver(observer);
        if (observerAdded) {
            this.wrapped.unregisterAdapterDataObserver(this.observer);
        }
    }

    private int getLoadingPosition() {
        return wrapped.getItemCount();
    }

    private boolean isLoadingPosition(int position) {
        return loading && position == getLoadingPosition();
    }

    public RecyclerView.LayoutManager getGridLayoutManager(Context context) {
        GridLayoutManager layout = new GridLayoutManager(context, 2);
        layout.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                if (isLoadingPosition(position)) {
                    return 2;
                } else {
                    return 1;
                }
            }
        });
        return layout;
    }

    public void updateLoading(boolean loading) {
        if (this.loading != loading) {
            this.loading = loading;
            try {
                if (loading) {
                    notifyItemInserted(getLoadingPosition());
                } else {
                    notifyItemRemoved(getLoadingPosition());
                }
            } catch (Throwable e) {
                // ignore
            }
        }
    }

    private class LoadingViewHolder extends RecyclerView.ViewHolder {

        LoadingViewHolder(ItemLoadingBinding binding) {
            super(binding.getRoot());
        }

    }
}

File: library/src/main/java/com/sprinklebit/library/presentation/common/adapter/LoadingAdapter.kt
package com.sprinklebit.library.presentation.common.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.sprinklebit.library.R
import timber.log.Timber

@Suppress("unused")
class LoadingAdapter(adapter: RecyclerView.Adapter<*>, private val threshold: Int = 5,
                     loadMoreListener: (() -> Any)? = null) : RecyclerViewAdapterWrapper(adapter) {

    private var loadMoreListener: LoadMoreListener? = null
    private var isHorizontal = false

    private var hasNextWasSet = false

    private var loading = true
    private var hasNext = true

    private val loadingPosition: Int
        get() = itemCount - 1

    companion object {
        const val LOADING_TYPE = -1023
    }

    init {
        loadMoreListener?.let {
            this.loadMoreListener = object : LoadMoreListener {
                override fun onLoadMore() {
                    it.invoke()
                }
            }
        }
    }

    override fun getAdapterDataObserver(): RecyclerView.AdapterDataObserver {
        return object : RecyclerView.AdapterDataObserver() {

            override fun onChanged() {
                notifyDataSetChanged()
            }

            override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
                notifyItemRangeChanged(positionStart, itemCount)
            }

            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                if (wrappedAdapter.itemCount == itemCount) {
                    if (hasNext) {
                        notifyItemRemoved(0)
                    }
                    notifyItemRangeInserted(positionStart, itemCount)
                    if (hasNext) {
                        notifyItemInserted(loadingPosition)
                    }
                } else {
                    notifyItemRangeInserted(positionStart, itemCount)
                }
            }

            override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                notifyItemRangeRemoved(positionStart, itemCount)
            }

            override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
                notifyItemMoved(fromPosition, toPosition)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == LOADING_TYPE) {
            val view = when (isHorizontal) {
                true -> LayoutInflater.from(parent.context).inflate(
                        R.layout.item_loading_horizontal, parent, false)
                false -> LayoutInflater.from(parent.context).inflate(
                        R.layout.item_loading_vertical, parent, false)
            }
            LoadingViewHolder(view)
        } else {
            wrappedAdapter.onCreateViewHolder(parent, viewType)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder !is LoadingViewHolder) {
            wrappedAdapter.onBindViewHolder(holder, position)
        }
        if (!loading && hasNext && loadingPosition - threshold <= position) {
            loadMore()
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isLoadingPosition(position)) {
            LOADING_TYPE
        } else {
            wrappedAdapter.getItemViewType(position)
        }
    }

    override fun getItemCount(): Int {
        val itemCount = wrappedAdapter.itemCount
        return if (hasNext) itemCount + 1 else itemCount
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        if (holder !is LoadingViewHolder) {
            wrappedAdapter.onViewRecycled(holder)
        }
    }

    private fun loadMore() {
        loading = true
        loadMoreListener?.onLoadMore()
    }

    private fun isLoadingPosition(position: Int): Boolean {
        return hasNext && position == loadingPosition
    }

    @Deprecated("Use constructor")
    fun setLoadMoreListener(loadMoreListener: LoadMoreListener) {
        this.loadMoreListener = loadMoreListener
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        isHorizontal = recyclerView.layoutManager!!.canScrollHorizontally()
    }

    @Deprecated("use setLoading and setHasNext")
    fun updateLoading(hasNext: Boolean) {
        try {
            if (hasNext) {
                resetLoading()
            } else {
                setFinished()
            }
        } catch (e: Throwable) {
            Timber.e(e)
        }
    }

    private fun setFinished() {
        if (!hasNext) {
            return
        }
        hasNext = false
        loading = false
        notifyItemRemoved(itemCount)
    }

    private fun resetLoading() {
        val add = !hasNext
        loading = false
        hasNext = true
        if (add) {
            notifyItemInserted(loadingPosition)
        } else {
            notifyItemChanged(loadingPosition)
        }
    }

    @Suppress("unused")
    fun setLoading(loading: Boolean) {
        this.loading = loading
    }

    @Suppress("unused")
    fun setHasNext(hasNext: Boolean) {
        if (!hasNextWasSet) loading = false
        hasNextWasSet = true

        try {
            if (hasNext) {
                val add = !this.hasNext
                this.hasNext = hasNext

                if (add) {
                    notifyItemInserted(loadingPosition)
                } else {
                    notifyItemChanged(loadingPosition)
                }
            } else {
                this.hasNext = hasNext
                notifyItemRemoved(loadingPosition)
            }
        } catch (e: Throwable) {
            Timber.e(e)
        }
    }

    @Suppress("unused")
    fun reset() {
        if (this.loading && hasNext && !hasNextWasSet) return

        val add = !this.hasNext

        loading = true
        hasNext = true
        hasNextWasSet = false

        if (add) {
            notifyItemInserted(loadingPosition)
        } else {
            notifyItemChanged(loadingPosition)
        }
    }

    interface LoadMoreListener {
        fun onLoadMore()
    }

    private inner class LoadingViewHolder internal constructor(itemView: View)
        : RecyclerView.ViewHolder(itemView)

}